app/models/document/paginated_history.rb

```diff
@@ -0,0 +1,56 @@
+class Document::PaginatedHistory
+  PER_PAGE = 30
+
+  attr_reader :document, :query
+
+  delegate :total_count, to: :query
+
+  def initialize(document, page)
+    @document = document
+    @query = document.edition_versions
+                     .includes(:user)
+                     .where.not(state: :superseded)
+                     .reorder(created_at: :desc, id: :desc)
+                     .page(page)
+                     .per(PER_PAGE)
+  end
+
+  def audit_trail
+    query.map.with_index do |version, index|
+      # [index - 1] returns next version, as array has newest first
+      AuditTrailEntry.new(version,
+                          is_first_edition: version.item_id == document.first_edition&.id,
+                          previous_version: query[index + 1])
+    end
+  end
+
+  class AuditTrailEntry
+    attr_reader :version, :is_first_edition, :preloaded_previous_version
```

Comment:
I think we could probably do without `preloaded_previous_version` being part of the public interface for this class.

In other words: I don't think this needs to be an `attr_reader` for `:preloaded_previous_version`. The only place we read its value is from within the class itself, so we could just as easily refer directly to the `@preloaded_previous_version` instance variable.

It'd also mean we can do without the associated test for that attribute. I think it's more of an internal implementation detail.

Reply:
In fact, I think we could also do without `is_first_edition` being on the public interface. That also looks like it's only used internally.

Reply:
Done and done. I had to get rid of the test for `is_first_edition` as well but this shouldn't matter because the behaviour is tested with the test of action determination, which is the only place the `is_first_edition` is used.